// seedling/index.js
var seedling = exports;
var mysql = require('mysql');
var db;


var connection = mysql.createConnection({
  host     : 'localhost',
  user     : 'root',
  password : '',
  database: "seederDATA"
});


//utility function to grab from the mongodb database and copy graph metadata into the mysql db
// this should not be used routinely and is for special purposes at the inset of development.

seedling.syncDBS = function(){

		db.createCollection('publicgraphs', function (err, collection) {
		
		    collection.find().toArray(function (err, items) {
		        items.forEach(function (obj) {
		            var graph = 
		            {
		            			graphid : obj.graphid, 
		            			userid : 1 , 
		            			graphname : obj.graphname, 
		            			description : obj.graphdesc, 
		            			nodecount : obj.graphmeta.nsize, 
		            			edgecount : obj.graphmeta.esize, 
		            			created : '2014-08-13 00:00:00'
		            }
		            var query = connection.query('INSERT INTO graphs SET ?;', graph, function(err, results) {
		            	});
		        });
		    });
		});

}





seedling.init = function(database_connection){
	
	//connect mysql database
	connection.connect();
	//hold reference to mongodb connection created by the server
	db = database_connection;
	//make sure graph data is synchronised between the two databases
	//this.syncDBS();

}
